### PR TITLE
Add Axvmdevice for vm

### PR DESCRIPTION
**Pull request by guoweikang**
Add Axvmdevice for vm

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+use alloc::string::String;
+use alloc::sync::Arc;
+use alloc::vec::Vec;
+use axdevice_base::BaseDeviceOps;
```
Comment (hky1999) [inline src/config.rs]: Add blank line between `use alloc::*` and `use ax*::*`

File: src/config.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    /// The IRQ (Interrupt Request) ID of the device.
+    pub irq_id: usize,
+    /// The type of emulated device.
+    pub emu_type: usize,
+    pub cfg_list: Vec<usize>,
+}
+
+/// Implementation of methods for EmuDeviceType.
+impl EmulatedDeviceConfig {
+    pub fn to_emu_dev(&self) -> AxResult<Arc<dyn BaseDeviceOps>> {
```
Comment (hky1999) [inline src/config.rs]: It looks so strange to impl this function to `EmulatedDeviceConfig`

File: src/device.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    pub fn new(config: AxVmDeviceConfig) -> Self {
+        let mut this = Self {
+            emu_devices: Vec::new(),
+        };
+
+        Self::init_emu_dev(&mut this, &config.emu_configs);
+        this
+    }
+
+    fn init_emu_dev(this: &mut Self, emu_configs: &Vec<EmulatedDeviceConfig>) {
```
Comment (hky1999) [inline src/device.rs]: init

File: src/device.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        for config in emu_configs {
+            let dev = config.to_emu_dev();
+            if let Ok(emu_dev) = dev {
+                this.emu_devices.push(emu_dev)
+            }
+        }
+    }
+
+    pub fn find_emu_dev(&self, ipa: usize) -> Option<Arc<dyn BaseDeviceOps>> {
```
Comment (hky1999) [inline src/device.rs]: find_dev

File: src/device.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        self.emu_devices
+            .iter()
+            .find(|&dev| dev.address_range().contains(ipa))
+            .cloned()
+    }
+
+    pub fn handle_emu_read(&self, addr: usize, width: usize) -> AxResult<usize> {
```
Comment (hky1999) [inline src/device.rs]: handle_mmio_read

File: src/device.rs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                "emu: {:?} handler read ipa {:#x}",
+                emu_dev.address_range(),
+                addr
+            );
+            return emu_dev.handle_read(addr, width);
+        }
+        panic!("emu_handle: no emul handler for data abort ipa {:#x}", addr);
+    }
+
+    pub fn handle_emu_write(&self, addr: usize, width: usize, val: usize) {
```
Comment (hky1999) [inline src/device.rs]: handle_mmio_write

File: src/lib.rs [hunk 7]
```diff
@@ -1,14 +1,12 @@
...
+#[macro_use]
+extern crate log;
 
-    #[test]
-    fn it_works() {
-        let result = add(2, 2);
-        assert_eq!(result, 4);
-    }
-}
+pub use config::EmulatedDeviceConfig;
```
Comment (hky1999) [inline src/lib.rs]: Mind the sequence of `mod::xxx` and `pub use xxx`.

General you should put `mod::xxx` together, then  `pub use xxx` after a blank line.

**Review by hky1999 — DISMISSED** [review]

File: src/config.rs (hunk 2 above)
Comment (buhenxihuan) [reply src/config.rs]: it is used for devices init

File: Cargo.toml [hunk 8]
```diff
@@ -3,4 +3,14 @@ name = "axdevice"
 version = "0.1.0"
 edition = "2021"
 
+[features]
+
 [dependencies]
+log = "=0.4.19"
+cfg-if = "1.0"
+serde = { version = "1.0.204", default-features = false, features = ["derive", "alloc"] }
+axdevice_base = { git = "https://github.com/arceos-hypervisor/axdevice_crates.git"}
```
Comment (hky1999) [inline Cargo.toml]: put this dep by  "System independent crates provided by ArceOS."

File: src/config.rs (hunk 2 above)
Comment (hky1999) [reply src/config.rs]: Such constructor should be implemented by xxx_device, through method like 
```Rust
impl VirtioBlock {
    fn init(config: EmulatedDeviceConfig) -> Self {}
}
```

**Comment by hky1999** [comment]
Also, pay attention to the ci errors.

File: src/device.rs [hunk 9]
```diff
@@ -0,0 +1,93 @@
+use alloc::sync::Arc;
+use alloc::vec::Vec;
+
+use crate::EmulatedDeviceConfig;
+use axdevice_base::{BaseDeviceOps, EmuDeviceType};
+
```
Comment (hky1999) [inline src/device.rs]: delete this blank line

File: src/device.rs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    /// Find specific device by ipa
+    pub fn find_dev(&self, ipa: GuestPhysAddr) -> Option<Arc<dyn BaseDeviceOps>> {
+        self.emu_devices
+            .iter()
+            .find(|&dev| dev.address_range().contains(ipa))
+            .cloned()
+    }
+
+    /// Handle the MMIO read
```
Comment (hky1999) [inline src/device.rs]: Explain this function with more details.

including arguments and the return value.

File: src/device.rs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        let mut this = Self {
+            emu_devices: Vec::new(),
+        };
+
+        Self::init(&mut this, &config.emu_configs);
+        this
+    }
+
+    /// According the emu_configs to init every  specific device
+    fn init(this: &mut Self, emu_configs: &Vec<EmulatedDeviceConfig>) {
```
Comment (hky1999) [inline src/device.rs:31]: Have you notice the awkward dependency logic here?

1. AxVMDevice provides the definition of `EmulatedDeviceConfig` here.
2. AxVMDevice has to depend on various of emulated devices.
3. emulated devices have to rely on `EmulatedDeviceConfig` for its constructor.

**Review by hky1999 — CHANGES_REQUESTED** [review]

File: src/device.rs [hunk 12]
```diff
@@ -0,0 +1,91 @@
+use crate::AxVmDeviceConfig;
+use alloc::sync::Arc;
+use alloc::vec::Vec;
+use axaddrspace::GuestPhysAddr;
```
Comment (hky1999) [inline src/device.rs:6]: add blank line between `use alloc::*` and use `axxx::*`

File: src/device.rs [hunk 13]
```diff
@@ -0,0 +1,91 @@
+use crate::AxVmDeviceConfig;
```
Comment (hky1999) [inline src/device.rs:1]: `use alloc::*` or `use core::*` should be placed in front of others.

File: src/lib.rs [hunk 14]
```diff
@@ -1,14 +1,24 @@
...
+//!
+//! The `alloc` crate is used to enable dynamic memory allocation in the absence of the standard library.
+//!
+//! The `log` crate is included for logging purposes, with macros being imported globally.
+//!
+//! The module is structured into two main parts: `config` and `device`, which manage the configuration and handling of AxVm devices respectively.
+//!
+//! Public exports include:
+//! - `AxVmDeviceConfig`: Configuration for AxVm devices.
+//! - `EmulatedDeviceConfig`: Configuration for emulated devices.
```
Comment (hky1999) [inline src/lib.rs]: Delete this from annotations

**Review by hky1999 — CHANGES_REQUESTED** [review]

**Comment by hky1999** [comment]
Also, always take care of CI errors.

File: src/device.rs (hunk 13 above)
Comment (buhenxihuan) [reply src/device.rs:1]: After running cargo fmt, it becomes like this

Comment (hky1999) [reply src/device.rs:1]: add blank lines between them

File: src/device.rs (hunk 9 above)
Comment (buhenxihuan) [reply src/device.rs]: 👌

**Review by hky1999 — APPROVED** [review]

**Pull request merged**